From 3e9c2127df2668005f668776cc0ad4b93cc3f21e Mon Sep 17 00:00:00 2001
From: Baokun Li <libaokun1@huawei.com>
Date: Tue, 18 Oct 2022 10:26:50 +0800
Subject: [PATCH 038/595] [Huawei] Revert "[Huawei] ext4: Fix warning in
 ext4_da_release_space"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186816

--------------------------------

This reverts commit 13ddcfaf80483a1c21cf6577387a23cd42ae1c56.

commit 13ddcfaf8048("[Huawei] ext4: Fix warning in ext4_da_release_space")
is a self-developed patch. The mainline patch (7f0d8e1d607c "ext4: fix
extent status tree race in writeback error recovery path") that fixes the
same issue has backported to this version(f17e61d9bee0). Therefore, the
self-developed patch can be reverted to avoid conflicts.

Signed-off-by: Baokun Li <libaokun1@huawei.com>
---
 fs/ext4/ext4.h           |  3 ---
 fs/ext4/extents_status.c |  5 -----
 fs/ext4/inode.c          | 11 ++---------
 fs/ext4/super.c          |  1 -
 4 files changed, 2 insertions(+), 18 deletions(-)

diff --git a/fs/ext4/ext4.h b/fs/ext4/ext4.h
index df43cda8fc2f..2373b4348f03 100644
--- a/fs/ext4/ext4.h
+++ b/fs/ext4/ext4.h
@@ -1165,9 +1165,6 @@ struct ext4_inode_info {
 	__u32 i_csum_seed;
 
 	kprojid_t i_projid;
-
-	/* Protect concurrent add cluster delayed block and remove block */
-	struct mutex i_clu_lock;
 };
 
 /*
diff --git a/fs/ext4/extents_status.c b/fs/ext4/extents_status.c
index dd679014db98..9a3a8996aacf 100644
--- a/fs/ext4/extents_status.c
+++ b/fs/ext4/extents_status.c
@@ -1433,7 +1433,6 @@ static int __es_remove_extent(struct inode *inode, ext4_lblk_t lblk,
 int ext4_es_remove_extent(struct inode *inode, ext4_lblk_t lblk,
 			  ext4_lblk_t len)
 {
-	struct ext4_sb_info *sbi = EXT4_SB(inode->i_sb);
 	ext4_lblk_t end;
 	int err = 0;
 	int reserved = 0;
@@ -1456,13 +1455,9 @@ int ext4_es_remove_extent(struct inode *inode, ext4_lblk_t lblk,
 	 * so that we are sure __es_shrink() is done with the inode before it
 	 * is reclaimed.
 	 */
-	if (sbi->s_cluster_ratio != 1)
-		mutex_lock(&EXT4_I(inode)->i_clu_lock);
 	write_lock(&EXT4_I(inode)->i_es_lock);
 	err = __es_remove_extent(inode, lblk, end, &reserved);
 	write_unlock(&EXT4_I(inode)->i_es_lock);
-	if (sbi->s_cluster_ratio != 1)
-		mutex_unlock(&EXT4_I(inode)->i_clu_lock);
 	ext4_es_print_tree(inode);
 	ext4_da_release_space(inode, reserved);
 	return err;
diff --git a/fs/ext4/inode.c b/fs/ext4/inode.c
index b8e02007b0c1..8c5c758033d5 100644
--- a/fs/ext4/inode.c
+++ b/fs/ext4/inode.c
@@ -1657,22 +1657,17 @@ static int ext4_insert_delayed_block(struct inode *inode, ext4_lblk_t lblk)
 			goto errout;
 		reserved = true;
 	} else {   /* bigalloc */
-		mutex_lock(&EXT4_I(inode)->i_clu_lock);
 		if (!ext4_es_scan_clu(inode, &ext4_es_is_delonly, lblk)) {
 			if (!ext4_es_scan_clu(inode,
 					      &ext4_es_is_mapped, lblk)) {
 				ret = ext4_clu_mapped(inode,
 						      EXT4_B2C(sbi, lblk));
-				if (ret < 0) {
-					mutex_unlock(&EXT4_I(inode)->i_clu_lock);
+				if (ret < 0)
 					goto errout;
-				}
 				if (ret == 0) {
 					ret = ext4_da_reserve_space(inode);
-					if (ret != 0) {   /* ENOSPC */
-						mutex_unlock(&EXT4_I(inode)->i_clu_lock);
+					if (ret != 0)   /* ENOSPC */
 						goto errout;
-					}
 					reserved = true;
 				} else {
 					allocated = true;
@@ -1684,8 +1679,6 @@ static int ext4_insert_delayed_block(struct inode *inode, ext4_lblk_t lblk)
 	}
 
 	ret = ext4_es_insert_delayed_block(inode, lblk, allocated);
-	if (sbi->s_cluster_ratio != 1)
-		mutex_unlock(&EXT4_I(inode)->i_clu_lock);
 	if (ret && reserved)
 		ext4_da_release_space(inode, 1);
 
diff --git a/fs/ext4/super.c b/fs/ext4/super.c
index 956827fd24ec..abf7ce63d0d0 100644
--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -1361,7 +1361,6 @@ static struct inode *ext4_alloc_inode(struct super_block *sb)
 	INIT_WORK(&ei->i_rsv_conversion_work, ext4_end_io_rsv_work);
 	ext4_fc_init_inode(&ei->vfs_inode);
 	mutex_init(&ei->i_fc_lock);
-	mutex_init(&ei->i_clu_lock);
 	return &ei->vfs_inode;
 }
 
-- 
2.17.1

